File: src/source/runtime/src/main/java/edu/uidaho/junicon/runtime/util/JarRunner.java
//========================================================================
// Copyright (c) 2015 Orielle, LLC.  
// All rights reserved.
//
// Redistribution and use in source and binary forms, with or without
// modification, are permitted provided that the following conditions
// are met:
//
// 1. Redistributions of source code must retain the above copyright
//    notice, this list of conditions and the following disclaimer.
//
// 2. Redistributions in binary form must reproduce the above copyright
//    notice, this list of conditions and the following disclaimer in the
//    documentation and/or other materials provided with the distribution.
//
// This software is provided by the copyright holders and contributors
// "as is" and any express or implied warranties, including, but not
// limited to, the implied warranties of merchantability and fitness for
// a particular purpose are disclaimed. In no event shall the copyright
// holder or contributors be liable for any direct, indirect, incidental,
// special, exemplary, or consequential damages (including, but not
// limited to, procurement of substitute goods or services; loss of use,
// data, or profits; or business interruption) however caused and on any
// theory of liability, whether in contract, strict liability, or tort
// (including negligence or otherwise) arising in any way out of the use
// of this software, even if advised of the possibility of such damage.
//========================================================================
package edu.uidaho.junicon.runtime.util;

import java.util.Arrays;
import java.util.jar.*;
import java.io.PrintStream;
import java.io.IOException;
import java.lang.reflect.Method;
import java.lang.reflect.InvocationTargetException;

/**
 * Runs the main-class in the given jar file's manifest.
 * Allows an executable jar file to be run with a classpath and dependencies.
 * This workaround is because "java -jar" ignores the user classpath.
 *
 * @author Peter Mills
 */
public class JarRunner {

  //====================================================================
  // Constructors
  //====================================================================
  public JarRunner () {
  }

  //====================================================================
  // Main.
  //====================================================================
  /**
   * Runs the main-class in the given jar file's manifest.
   * If no manifest or main-class is found, uses the tail of the jarname
   * without its dot-extension as the classname.
   * If classname is "manifest" or omitted, JarRunner uses manifest as above,
   * otherwise overrides to run that class instead.
   * <BR>
   * Usage: edu.uidaho.junicon.runtime.util.JarRunner
   *		jarfile [classname|"manifest"] [args]
   */
  public static void main (String[] args) {
	String usage = "Usage: JarRunner jarfile [classname|manifest] [args]"
	    + "\n\t"
	    + "Runs the main-class in the given jar file's manifest.";

	// Process command-line arguments
	if ((args.length < 1) || (args[0] == null)) {
		exit(usage);
	}
	String jarname = args[0];
	String classname = null;
	String[] mainArgs = new String[0];
	if (args.length > 1) {
		classname = args[1];
		if (args.length > 2) {
		    mainArgs = Arrays.copyOfRange(args, 2, args.length);
		}
	}

	// Get classname from manifest if "manifest" or omitted
	if ((classname == null) || classname.equals("manifest")) {
	  String tail = getFilenameTail(jarname);
	  classname = getFilenameWithoutExtension(tail);

	  // Get class with main method from jar manifest, if it exists
	  Manifest manifest = null;
	  try {
		JarFile jar = new JarFile(jarname);
		manifest = jar.getManifest();
	  } catch (IOException e) {
		exit("Error loading jar file: " + jarname, e);
	  }
	  if (manifest != null) {
	    Attributes attributes = manifest.getMainAttributes();
	    String mainClass = attributes.getValue(Attributes.Name.MAIN_CLASS);
	    if (mainClass != null) {
		classname = mainClass;
		// fatal("Jar file mainfest does not contain a Main-Class");
	    }
	  }
	}

	// Invoke main-class with arguments
	try {
		Class<?> clazz = Class.forName(classname);
		Method method = clazz.getMethod("main", String[].class);
		method.invoke(null, (Object) mainArgs);
	} catch (ClassNotFoundException e) {
		exit("Class not found: " + classname);
	} catch (NoSuchMethodException e) {
		exit("Class does not define a main method: " + classname);
	} catch (IllegalAccessException e) {
		exit("", e.getCause());
        } catch (InvocationTargetException e) {
		exit("", e.getCause());
        }
  }

  //====================================================================
  // Utilities.
  //====================================================================

  /**
   * Gets head of filename, i.e., prefix before first dot.
   * Strips path up to "/" or "\".
   */
  public static String getFilenameTail (String filename) {
	if ((filename == null) || filename.isEmpty()) { return filename; }
	int slash = filename.lastIndexOf("/");
	int backslash = filename.lastIndexOf("\\");
	if ((slash >= 0) || (backslash >= 0)) {
		int tailpos = ((slash > backslash) ? slash : backslash) + 1;
		if (tailpos > filename.length()) {
			filename = filename.substring(tailpos);
		}
	}
	return filename;
  }

  /**
   * Get filename extension, i.e., text after last dot.
   * Returns extension, or empty if no extension, or null if filename is null.
   */
  public static String getFilenameExtension (String filename) {
	if (filename == null) { return null; }
	int pos = filename.lastIndexOf(".");
	if ((pos < 0) || (pos >= filename.length())) { return ""; }
	return filename.substring(pos+1);
  }

  /**
   * Get filename root, i.e., text before last dot.
   */
  public static String getFilenameWithoutExtension (String filename) {
	if (filename == null) { return null; }
	int pos = filename.lastIndexOf(".");
	if (pos < 0) { return filename; }
	return filename.substring(0, pos);
  }

  //====================================================================
  // Logger.
  //====================================================================
  private static PrintStream err = System.err;

  static void fatal (String message) {
        err.println(message);
  }

  static void fatal (String message, Throwable e) {
        err.println(message);
	e.printStackTrace();
  }

  static void exit (String message) {
	fatal(message);
        System.exit(1);
  }

  static void exit (String message, Throwable e) {
	fatal(message, e);
        System.exit(1);
  }

}

//==== END OF FILE

File: src/source/interpreter/src/main/resources/config/bin/windows/JarRunner.cpp
//============================================================================
// Copyright (c) 2015 Orielle, LLC.  
// All rights reserved.
//
// Redistribution and use in source and binary forms, with or without
// modification, are permitted provided that the following conditions are met:
//
// 1. Redistributions of source code must retain the above copyright
//    notice, this list of conditions and the following disclaimer.
//
// 2. Redistributions in binary form must reproduce the above copyright
//    notice, this list of conditions and the following disclaimer in the
//    documentation and/or other materials provided with the distribution.
//
// This software is provided by the copyright holders and contributors
// "as is" and any express or implied warranties, including, but not
// limited to, the implied warranties of merchantability and fitness for
// a particular purpose are disclaimed. In no event shall the copyright
// holder or contributors be liable for any direct, indirect, incidental,
// special, exemplary, or consequential damages (including, but not
// limited to, procurement of substitute goods or services; loss of use,
// data, or profits; or business interruption) however caused and on any
// theory of liability, whether in contract, strict liability, or tort
// (including negligence or otherwise) arising in any way out of the use
// of this software, even if advised of the possibility of such damage.
//============================================================================
#include <stdlib.h>	/* system, NULL, EXIT_FAILURE */
#include <stdio.h>	/* printf */
#include <string.h>
#include <process.h>
#include <iostream>	/* cout, cin, cerr */

/**
 * Execute attached Java jar file.
 *
 * This program runs java on the main-class in the appended jar file.
 * To create an application, the jar file should be appended to this executable,
 * and its manifest should hold the main-class to be executed.
 * The concatenated file can double both as a jar/zip and as an executable,
 * because zip files read from the end, while executables read from the front.
 *
 * If JAVA_HOME is defined, uses that for java; otherwise searches the PATH.
 * If CLASSPATH is defined, adds this jarfile to the classpath, and then
 *	invokes JarRunner to reflectively invoke the main-class in the manifest.
 *	This workaround is because "java -jar" ignores the user classpath.
 * Otherwise, uses "java -jar" to invoke the self-contained executable jar.
 *
 * Sets the environment variable JARRUNNER_ARG0 to this executable filename,
 * and also sets JARRUNNER_ISWINDOWS=true.
 * The executable filename will have .exe appended if not there,
 * so that the original jar file name is available within the Java application.
 * This workaround is because arg0 is dropped from Java's main(args).
 *
 * Compile under Visual Studio C++ with:
 *	vcvars32.bat
 *	cl.exe /EHsc JarRunner.cpp
 * Compile under mingw with:
 *	g++ JarRunner.cpp -static -o jarrunner.exe
 */
int main (int argc, char* argv[], char* envp[]) {
  // First option is name of process: {"java", "java", "-jar", "this", NULL}
  int optionsOffset = 3;	// Start of command-line options, if java -jar
  int maxOptionsOffset = 6;	// Start of command-line options, if JarRunner
  char *javaHome;		// From environment
  char *classpath;
  char *thisFileName;		// Arg0 to this process
  char *command;		// Invoked command
  char **mainArgs;		// Invoked arguments
  char **mainEnv;
  std::string arg0;		// Quoted argument 0 to pass to invoked command
  std::string thisExe;		// Full name of this invoked file, with .exe
  std::string javaPath;
  std::string jarClasspath;
  bool isJavaHome = false;
  int len;

  mainArgs = new char*[argc + maxOptionsOffset];

  // Get invoked file name.  Append .exe if not there, for correct jar file name
  thisFileName = argv[0];
  if (thisFileName == NULL) { thisFileName = (char *) ""; }
  thisExe = thisFileName;
  std::string suffix = ".exe";
  if (! ((thisExe.length() >= suffix.length()) &&
        	(0 == thisExe.compare (thisExe.length() - suffix.length(),
			suffix.length(), suffix)))) {
	thisExe += ".exe";
  }

  // Set process to execute
  command = (char *) "java.exe";
  arg0 = command;

  // Set java program location from JAVA_HOME, if defined
  // if getenv(JAVA_HOME), set mainArgs[0] to it/bin/java, just do execv.
  javaHome = getenv("JAVA_HOME");
  if ((javaHome != NULL) && ((len = strlen(javaHome)) > 0)) {
	javaPath = javaHome;
	if ((javaHome[len-1] == '/') || (javaHome[len-1] == '\\')) {
		javaPath += "bin/java.exe";
	} else {
		javaPath += "/bin/java.exe";
	}
	command = (char *) javaPath.c_str();
	arg0 = "\"" + javaPath + "\"";
	isJavaHome = true;
  }

  mainArgs[0] = (char *) arg0.c_str(); // Name of process
  mainArgs[1] = (char *) "-jar"; // First option. JarRunner if CLASSPATH defined
  mainArgs[2] = (char *) thisExe.c_str(); // Name of file with jar is this

  // Decide if use "-cp classpath JarRunner" instead of "-jar"
  // if getenv(CLASSPATH), use JarRunner, in case have runnable there.
  classpath = getenv("CLASSPATH");
  if ((classpath != NULL) && ((len = strlen(classpath)) > 0)) {
	jarClasspath = classpath;
	jarClasspath += ";";
	jarClasspath += thisExe;
	optionsOffset = maxOptionsOffset;
	mainArgs[1] = (char *) "-cp";
	mainArgs[2] = (char *) jarClasspath.c_str();
	mainArgs[3] = (char *) "edu.uidaho.junicon.runtime.util.JarRunner";
	mainArgs[4] = (char *) thisExe.c_str();
	mainArgs[5] = (char *) "manifest";	// JarRunner uses manifest
  }

  // Copy in other options argv[1..end].  Must quote strings, if have spaces.
  std::string *quoted;
  quoted = new std::string[argc];
  for (int i=1; i<argc; i++) {		// argv[1] --> mainArgs[optionsOffset]
	quoted[i] = argv[i];
	if (quoted[i].find(' ')) {
		quoted[i] = "\"" + quoted[i] + "\"";
	}
	mainArgs[(i-1) + optionsOffset] = (char *) quoted[i].c_str();
  }

  // End mainArgs with NULL
  mainArgs[(argc-1) + optionsOffset] = NULL;

  // Copy over environment, add JARRUNNER_ARG0, JARRUNNER_ISWINDOWS
  int envc = 0;
  if (envp != NULL) {
	while (envp[envc] != NULL) { envc++; }
  }
  mainEnv = new char*[envc + 3];
  for (int i=0; i < envc; i++) {
	mainEnv[i] = envp[i];
  }
  std::string env_arg0("JARRUNNER_ARG0");
  env_arg0 += "=";
  env_arg0 += thisExe;
  mainEnv[envc] = (char *) env_arg0.c_str();
  mainEnv[envc + 1] = (char *) "JARRUNNER_ISWINDOWS=true";
  mainEnv[envc + 2] = NULL;

  // Debug
  /****
  std::cerr << "Arguments" << "\n";
  for (int i=0; (mainArgs[i] != NULL); i++) {
    std::cerr << mainArgs[i] << "\n";
  }
  std::cerr << "End of arguments" << "\n";
  std::cerr << "Using PATH " << (! isJavaHome) << "\n";

  std::cerr << "Environment" << "\n";
  for (int i=0; (mainEnv[i] != NULL); i++) {
    std::cerr << mainEnv[i] << "\n";
  }
  std::cerr << "End of environment" << "\n";
  ****/

  // Invoke java on this jar file
  if (isJavaHome) {
#ifdef _P_WAIT
	_spawnve(_P_WAIT, command, (char **) mainArgs, (char **) mainEnv);
#else
	spawnve(P_WAIT, command, (char **) mainArgs, (char **) mainEnv);
#endif
  } else {
#ifdef _P_WAIT
	_spawnvpe(_P_WAIT, command, (char **) mainArgs, (char **) mainEnv);
#else
	spawnvpe(P_WAIT, command, (char **) mainArgs, (char **) mainEnv);
#endif
  }
  return 0;
}

//====
// // P_WAIT, P_NOWAIT, P_OVERLAY, P_DETACH (Windows has _ prefix)
// int pid = _spawnvpe(_P_NOWAIT, command, args, env)
// _cwait(&termstat_int, pid, _WAIT_CHILD)
//====

//==== END OF FILE
